sunpy/physics/differential_rotation.py

```diff
@@ -227,102 +409,374 @@ def _warp_sun_coordinates(xy, smap, dt: u.s, **diffrot_kwargs):
-            smap = sunpy.map.Map(smap_data, smap_meta)
-
-    warp_args = {'smap': smap, 'dt': dt}
-    warp_args.update(diffrot_kwargs)
+    # Create the arguments for the warp function.
+    warp_args = {'smap': smap, 'new_observer': new_observer}
+    warp_args.update(diff_rot_kwargs)
+
     # Apply solar differential rotation as a scikit-image warp
-    out = transform.warp(to_norm(smap_data), inverse_map=_warp_sun_coordinates,
-                         map_args=warp_args)
+    out_data = transform.warp(to_norm(smap_data), inverse_map=_warp_sun_coordinates, map_args=warp_args)
 
     # Recover the original intensity range.
-    out = un_norm(out, smap.data)
+    out_data = un_norm(out_data, smap.data)
 
-    # Update the meta information with the new date and time, and reference pixel.
+    # Update the meta information with the new date and time.
     out_meta = deepcopy(smap.meta)
     if out_meta.get('date_obs', False):
         del out_meta['date_obs']
-    out_meta['date-obs'] = "{}".format(new_time.strftime('%Y-%m-%dT%H:%M:%S'))
+    out_meta['date-obs'] = new_observer.obstime.strftime("%Y-%m-%dT%H:%M:%S.%f")
+
+    # Need to update the observer location for the output map.
+    out_meta['hglt_obs'] = new_observer.lat.value
+    out_meta['hgln_obs'] = new_observer.lon.value
+    out_meta['dsun_obs'] = new_observer.radius.to(u.m).value
+
+    if is_sub_full_disk:
+        # Define a new reference pixel and the value at the reference pixel.
+        # Note that according to the FITS convention the first pixel in the
+        # image is at (1.0, 1.0).
+        center_rotated = solar_rotate_coordinate(smap.center, observer=new_observer, **diff_rot_kwargs)
+        out_meta['crval1'] = center_rotated.Tx.value
+        out_meta['crval2'] = center_rotated.Ty.value
+        out_meta['crpix1'] = 1 + smap.data.shape[1]/2.0 + ((center_rotated.Tx - smap.center.Tx)/smap.scale.axis1).value
+        out_meta['crpix2'] = 1 + smap.data.shape[0]/2.0 + ((center_rotated.Ty - smap.center.Ty)/smap.scale.axis2).value
+        return sunpy.map.Map(out_data, out_meta).submap(rotated_bl, rotated_tr)
+    else:
+        return sunpy.map.Map(out_data, out_meta)
+
+
+# Functions that calculate useful quantities from maps.
+def all_pixel_indices_from_map(smap):
+    """
+    Returns pixel pair indices of every pixel in a map.
+
+    Parameters
+    ----------
+    smap : `~sunpy.map.Map`
+        A SunPy map.
+
+    Returns
+    -------
+    out : `~numpy.array`
+        A numpy array with the all the pixel indices built from the
+        dimensions of the map.
+    """
+    return np.meshgrid(*[np.arange(v.value) for v in smap.dimensions]) * u.pix
 
-    if submap:
-        crval_rotated = solar_rotate_coordinate(smap.reference_coordinate, new_time, **diffrot_kwargs)
-        out_meta['crval1'] = crval_rotated.Tx.value
-        out_meta['crval2'] = crval_rotated.Ty.value
 
-    return sunpy.map.Map((out, out_meta))
+def all_coordinates_from_map(smap):
```

Comment:
This should be moved onto `GenericMap` with an API identical to [NDCube.axis_world_coordinates](https://github.com/sunpy/ndcube/blob/master/ndcube/ndcube.py#L300) although we have a much simiplier case for Map.

Reply:
Not sure I should change this API so it looks like something similar to something that is not in SunPy yet.

Reply:
The long term plan is to have Map be based on NDCube, so it seems sensible to me to implement this on map now, and maybe have a couple of small behavioural tweaks in the future rather than at some point in the future move this API into a method of map and have to deprecate this one?